internal/docstore/drivertest/drivertest.go

```diff
@@ -0,0 +1,181 @@
+// driver.
+package drivertest // import "gocloud.dev/internal/docstore/drivertest"
+
+import (
+	"context"
+	//	"math"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	ds "gocloud.dev/internal/docstore"
+	"gocloud.dev/internal/docstore/driver"
+)
+
+// Harness descibes the functionality test harnesses must provide to run
+// conformance tests.
+type Harness interface {
+	// MakeCollection makes a driver.Collection for testing.
+	MakeCollection(context.Context) (driver.Collection, error)
+
+	// Close closes resources used by the harness.
+	Close()
+}
+
+// HarnessMaker describes functions that construct a harness for running tests.
+// It is called exactly once per test; Harness.Close() will be called when the test is complete.
+type HarnessMaker func(ctx context.Context, t *testing.T) (Harness, error)
+
+// RunConformanceTests runs conformance tests for provider implementations of docstore.
+func RunConformanceTests(t *testing.T, newHarness HarnessMaker) {
+	t.Run("Create", func(t *testing.T) { withCollection(t, newHarness, testCreate) })
+	t.Run("Put", func(t *testing.T) { withCollection(t, newHarness, testPut) })
+	t.Run("Replace", func(t *testing.T) { withCollection(t, newHarness, testReplace) })
+	t.Run("Get", func(t *testing.T) { withCollection(t, newHarness, testGet) })
+	t.Run("Delete", func(t *testing.T) { withCollection(t, newHarness, testDelete) })
+	t.Run("Update", func(t *testing.T) { withCollection(t, newHarness, testUpdate) })
+}
+
+const KeyField = "_id"
+
+func withCollection(t *testing.T, newHarness HarnessMaker, f func(*testing.T, *ds.Collection)) {
+	ctx := context.Background()
+	h, err := newHarness(ctx, t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer h.Close()
+
+	dc, err := h.MakeCollection(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	coll := ds.NewCollection(dc)
+	f(t, coll)
+}
+
+type docmap = map[string]interface{}
+
+var nonexistentDoc = docmap{KeyField: "doesNotExist"}
+
+func testCreate(t *testing.T, coll *ds.Collection) {
+	named := docmap{KeyField: "testCreate1", "b": true}
+	unnamed := docmap{"b": false}
+	// Attempt to clean up
+	defer func() {
+		_, _ = coll.Actions().Delete(named).Delete(unnamed).Do(context.Background())
+	}()
+
+	mustRunAndCompare(t, coll, named, coll.Actions().Create(named))
+	mustRunAndCompare(t, coll, unnamed, coll.Actions().Create(unnamed))
+	// Can't create an existing doc.
+	shouldFail(t, coll.Actions().Create(named))
+}
+
+func testPut(t *testing.T, coll *ds.Collection) {
+	named := docmap{KeyField: "testPut1", "b": true}
+	mustRunAndCompare(t, coll, named, coll.Actions().Put(named)) // create new
+	named["b"] = false
+	mustRunAndCompare(t, coll, named, coll.Actions().Put(named)) // replace existing
+}
+
+func testReplace(t *testing.T, coll *ds.Collection) {
+	doc1 := docmap{KeyField: "testReplace", "s": "a"}
+	mustRun(t, coll.Actions().Put(doc1))
+
+	doc1["s"] = "b"
+	mustRunAndCompare(t, coll, doc1, coll.Actions().Replace(doc1))
+
+	// Can't replace a nonexistent doc.
```

Comment:
This doesn't seem consistent with not getting an error from deleting a non-existent doc.

Reply:
Replace is defined to mean "insert this doc if it exists".

Put, on the other hand, will write the doc either way, so that _is_ consistent with delete.

I justify the semantics of delete near where I define it.